javascript/ql/src/semmle/javascript/security/performance/SuperlinearBackTracking.qll

```diff
@@ -0,0 +1,141 @@
+/**
+ * Provides classes for working with regular expressions that can
+ * perform backtracking in superlinear time.
+ */
+
+import javascript
+
+/**
+ * A regular expression term that permits unlimited repetitions.
+ */
+private class InfiniteRepetitionQuantifier extends RegExpQuantifier {
+  InfiniteRepetitionQuantifier() {
+    this instanceof RegExpPlus
+    or
+    this instanceof RegExpStar
+    or
+    this instanceof RegExpRange and not exists(this.(RegExpRange).getUpperBound())
+  }
+}
+
+/**
+ * Holds if `t` matches at least an epsilon symbol.
+ *
+ * That is, this term does not restrict the language of the enclosing regular expression.
+ *
+ * This is implemented as an under-approximation, and this predicate does not hold for sub-patterns in particular.
+ */
+private predicate matchesEpsilon(RegExpTerm t) {
+  t instanceof RegExpStar
+  or
+  t instanceof RegExpOpt
+  or
+  t.(RegExpRange).getLowerBound() = 0
+  or
+  exists(RegExpTerm child |
+    child = t.getAChild() and
+    matchesEpsilon(child)
+  |
+    t instanceof RegExpAlt or
+    t instanceof RegExpGroup or
+    t instanceof RegExpPlus or
+    t instanceof RegExpRange
+  )
+  or
+  matchesEpsilon(t.(RegExpBackRef).getGroup())
+  or
+  forex(RegExpTerm child | child = t.(RegExpSequence).getAChild() | matchesEpsilon(child))
+}
+
+/**
+ * Gets a term that matches the symbol immediately before `t` is done matching.
```

Comment:
Could you rephrase this? I don't understand what this means, and I can't figure it out from the implementation.

Reply:
I have tried that already several times, I think I stuck in the wrong corner. Let me state the *purpose* of this predicate, and then you may view this problem from the right angle to help me get a better naming/docstring.

The predicate is used to find a term `t1`: 
1. `t1` consumes the last symbol just before an infinitely repeating term `t2`. 
   - for example: `t1` is `a`, and `t2` is `b+` here: `/...ab+.../` and here: `/...a(foo)?b+.../`. (this is what `getAMatchPredecessor(this.getPredecessor())` implements)
2. `t1` is infinitely repeating (this is what `InfiniteRepetitionQuantifier` implements)
3. `t1` can consume at least one of the same symbols as `t2` (this is what `compatible` implements)

Reply:
Yeah I can see how it's not easy to capture in a short sentence.

I might be enough to just add some examples, like:
- For `b?` in `ab?c` this gets `a` and `b` (in addition to `b?` itself).
- For `(ab|cd)` this gets `b` and `d` (in addition to `ab|cd` and `(ab|cd)`).